### PR TITLE
Refactor Che installation

### PR DESCRIPTION
**Pull request by xcoulon**
- simplify controller logic when checking for resources before creating them
- add log messages for each case
- refactor unit tests, grouping by "level" of resource to process

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 1]
```diff
@@ -120,109 +119,97 @@ func (r *ReconcileCheInstallation) Reconcile(request reconcile.Request) (reconci
...
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: cheOpNamespace}, ns); err != nil {
-		if errors.IsNotFound(err) {
-			logger.Info("Creating a namespace for che operator", "Namespace", cheOpNamespace)
-			namespace := NewNamespace(cheOpNamespace)
-			if err := controllerutil.SetControllerReference(cheInstallation, namespace, r.scheme); err != nil {
+	namespace := NewNamespace(cheOpNamespace)
+	if err := controllerutil.SetControllerReference(cheInstallation, namespace, r.scheme); err != nil {
+		return false, err
+	}
+	if err := r.client.Create(context.TODO(), namespace); err != nil {
```
Comment (nurali-techie) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]: To me this seems to be an anti-pattern "always create" against current way "create if not exists".

Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: I understand your concern, but I assume the checking is done by the server, so there is no creation if a resource with the same name already exists.
The purpose of these functions are more straightforward to understand: they attempt to create the resource and treat the error accordingly. As a result, the complexity of the code is reduced. Also, I believe that trying to create immediately is more "atomic" in the sense that checking if the resource exists is done by the server.

But sure, it's a shift in term of approach...

File: go.mod298498081.tmp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+module github.com/codeready-toolchain/toolchain-operator
```
Comment (alexeykazakov) [inline go.mod298498081.tmp]: leftover

File: go.mod [hunk 3]
```diff
@@ -1,64 +1,83 @@
...
 	k8s.io/gengo v0.0.0-20191120174120-e74f70b9b27e // indirect
+	k8s.io/klog v1.0.0 // indirect
 	k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
 	k8s.io/utils v0.0.0-20191114184206-e782cd3c129f // indirect
 	sigs.k8s.io/controller-runtime v0.2.2
 	sigs.k8s.io/controller-tools v0.2.4
+
 )
 
 // Pinned to kubernetes-1.14.1
```
Comment (alexeykazakov) [inline go.mod]: It doesn't seem to be kube 1.14.1 anymore. Some of libs were upgraded and some were downgraded...

**Pull request merged**